text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+    <Thing />
+  {{else}}
+    <OtherThing />
+  {{/if}}
+
+  {{! ⬆️compiles to ⬇️ }}
+  <Thing />
+```
+
+### Handlebars Macro: macroDependencySatisfies
+
+Acts like a helper that returns a boolean. Like the `dependencySatisfies` JS macro, it can only resolve things that are **allowed dependencies**, so the same need for peer dependencies and/or **Optional Peer Dependencies** applies.
+
+```hbs
+<SomeComponent @canAnimate={{macroDependencySatisfies "liquid-fire" "*"}} />
+{{! ⬆️compiles to ⬇️ }}
+<SomeComponent @canAnimate={{true}} />
+```
+
+### Handlebars Macro: macroMaybeAttrs
+
+There is one place where `{{#if}}` doesn't work: within "element space". If you want to _sometimes_ set an attribute, but sometimes not, this doesn't work:
+
+```hbs
+<div {{#if this.testing}} data-test-target={{@id}} {{/if}} />
+```
+
+`macroMaybeAttrs` exists to conditionally compile away attributes and arguments out of element space:
+
+```hbs
+<div {{macroMaybeAttrs (getConfig "ember-test-selectors" "enabled") data-test-target=@id }} />
+```
+
+It can be placed on both HTML elements and angle bracket component invocations.
+
+### Handlebars Macro: macroFailBuild
+
+Like the JS `failBuild` macro.
+
+```hbs
+{{#if (macroCondition (dependencySatisfies "some-peer-dep" "^3.0.0")) }}
+  <UseTheThing />
+{{else}}
+  {{macroFailBuild "You tried to use <MyFancyComponent/> but it requires some-peer-dep ^3.0.0"}}
+{{/if}}
+```
+
+### Macros: Overall Design
+
+All the macros are intended to be valid syntax. They shouldn't break parsing or linting.
+
+While we guarantee that branch elimination will run in production builds, we _don't_ guarantee that in development. The macros are designed so that in development they may have _runtime_ implementations. This is powerful because it lets us produce a single build that works in multiple contexts. For example:
+
+- it solves the longstanding problem that when you run your tests by visiting `localhost:4200/tests` the tests see the `development` environment, not the `test` environment. To get the test environment you can't use `ember serve`, you must use `ember test`. This has remained unfixed because it's expensive to do the whole build twice for the two environments.
+
+  We can solve this problem by producing a _single_ build containing _both_ environments, guarded by the macro system. The macros can evaluate at runtime, allowing each environment to get the right thing. In production builds, test-only or dev-only branches will still be eliminated.
+
+- it makes Fastboot builds simpler because we can guard the fastboot-only and browser-only code with the macro system. In development, we can run a single build that leaves both branches in and evaluates the macros at runtime.
+
+The macros package (`@ember/macros` as proposed, `@embroider/macros` as implemented) will work in both regular ember-cli and in Embroider. And it will work in both V1 and V2 packages.
+
+## Peer Dependencies
```

Comment:
Are you debating failing the build when a peer dep isn't satisfied? I always imagine the following scenario:

package A has a peer dep on package B
package C has a direct dep on package B
App includes package A, then indirectly satisdfies the peer dep by including package C (on purpose, package C is a bundling package)

npm will warn like crazy, even though the app consumer may know what they are doing. I wouldn't want a strict implementation of peer deps to fail in this case.

Reply:
This RFC doesn't fail the build for a missing peerDep.

But if you actually try to resolve a peerDep within your Ember code, Embroider implements that resolution strictly. You won't find an indirectly satisfied peerDep because that's just not safe to rely on. It happens to work often based on hoisting, but it also fails sometimes, and I have seen it fail many times in the wild.

Plus, you'll have a bad time adopting Yarn PnP or some other future more optimizable package system if you are relying on this kind of implicit dependencies.

So embroider is strict. A v2 package can't resolve things that aren't **allowed dependencies**, as defined in the definitions section.

The solution to your example is that package C should have a peerDep on B, so that B can be added once to the app and shared by both A and C (and the app). 

"bundling package" is a common pattern today in Ember addons, but for historical and no-longer-true-reasons. Now that importing libraries directly from NPM is easy, bundling packages either go away (if they were only bundling) or become separate libraries (if they were providing some ember-specific things like components and helpers that work nicely with the library).